### DISCUSSION TITLE
Create proper change log data when updating objects in custom scripts with NetBox 2.11

### DISCUSSION THREAD

**Discussion by netsandbox** [netbox-community/netbox#6491] [Q&A] [answered]
Hi,

in NetBox 2.10.10 the following worked in a custom script and created proper change log records:

```python
ip_address = IPAddress.objects.get(address="192.168.178.10/24")
ip_address.dns_name = "example.com"
ip_address.save()
```

In NetBox 2.11.4 the above creates a change log record without **Pre-Change Data** and **Difference**.
I always have to create a snapshot after getting the object to have proper change log records:

```python
ip_address = IPAddress.objects.get(address="192.168.178.10/24")
ip_address.snapshot()
ip_address.dns_name = "example.com"
ip_address.save()
```
Is this the correct way?

9c967ee3 created a `get_object_with_snapshot` method on `ModelViewSet`.
Maybe it makes sense to add such method also to `ChangeLoggingMixin`so this would work:

```python
ip_address = IPAddress.objects.get_object_with_snapshot(address="192.168.178.10/24")
ip_address.dns_name = "example.com"
ip_address.save()
```

And this should be also documented at https://netbox.readthedocs.io/en/stable/additional-features/custom-scripts/.

In some custom scripts I use the `update_or_create` method (e.g. VirtualMachine.objects.update_or_create).
This method also never takes a snapshot so if I update a object with this method the change log records also never have **Pre-Change Data** and **Difference**.
Do I now have to rewrite my custom scripts to separate object create and object update, so that I can create a snapshot before updating an object to have proper change log records?

**Comment by jeremystretch** [comment, marked as answer]
Snapshots aren't taken automatically because doing so effectively saves a copy of the object on itself. This is obviously wasteful if scenarios where there's no intent to modify and save the object.

> Is this the correct way?

Yes, calling `snapshot()` on the instance will capture the pre-change data.

> Maybe it makes sense to add such method also to ChangeLoggingMixin

It's probably worth discussing further as a feature request.

> Do I now have to rewrite my custom scripts to separate object create and object update

Yes, because it's not possible to capture the pre-change state of an object when writing to the database directly.
